File: triplate-ts/src/parser.ts
import { TriplateSyntaxError } from './errors.js';
import type {
  CompiledTemplateData,
  Cond,
  ExampleSet,
  ForNode,
  IfNode,
  Node,
  RefPath,
  Schema,
  TypeExpr,
} from './ast.js';
import { tokenize, type Token } from './lexer.js';

export function parse(template: string): CompiledTemplateData {
  const tokens = tokenize(template);

  let schema: Schema | undefined;
  const examples: ExampleSet[] = [];
  const body: Token[] = [];
  for (const t of tokens) {
    if (t.kind === 'params') {
      if (schema) throw new TriplateSyntaxError('duplicate params section in frontmatter', t.line, t.column);
      const byName: Record<string, TypeExpr> = {};
      for (const d of t.decls) {
        if (byName[d.name]) throw new TriplateSyntaxError(`duplicate parameter: ${d.name}`, t.line, t.column);
        byName[d.name] = d.type;
      }
      schema = { params: t.decls, byName };
    } else if (t.kind === 'examples') {
      if (examples.some((e) => e.id === t.id)) {
        throw new TriplateSyntaxError(`duplicate example id: ${t.id}`, t.line, t.column);
      }
      examples.push({ id: t.id, description: t.description, bindings: t.bindings });
    } else if (t.kind === 'text' && schema === undefined) {
      if (t.value.trim() !== '') throw new TriplateSyntaxError('content before the --- frontmatter header');
    } else {
      body.push(t);
    }
  }
  if (!schema) throw new TriplateSyntaxError('missing --- frontmatter header (--- params { … } ---)');

  const tree = buildTree(body);
  validateScopes(tree, new Set(Object.keys(schema.byName)));
  return { schema, examples, body: tree };
}

type Frame =
  | { kind: 'root'; children: Node[] }
  | { kind: 'for'; node: ForNode }
  | { kind: 'if'; node: IfNode; inElse: boolean };

function buildTree(tokens: Token[]): Node[] {
  const root: Node[] = [];
  const stack: Frame[] = [{ kind: 'root', children: root }];
  const current = (): Node[] => {
    const f = stack[stack.length - 1];
    if (f.kind === 'root') return f.children;
    if (f.kind === 'for') return f.node.body;
    return f.inElse ? f.node.elseBody! : f.node.branches[f.node.branches.length - 1].body;
  };
  const append = (n: Node) => current().push(n);

  for (const t of tokens) {
    switch (t.kind) {
      case 'text':
        append({ type: 'text', value: t.value });
        break;
      case 'value':
        append({ type: 'value', path: t.path, line: t.line, column: t.column });
        break;
      case 'interp':
        append({ type: 'interp', parts: t.parts, lang: t.lang, datatype: t.datatype, line: t.line, column: t.column });
        break;
      case 'iri':
        append({ type: 'iri', parts: t.parts, line: t.line, column: t.column });
        break;
      case 'for': {
        const node: ForNode = {
          type: 'for', item: t.item, source: t.source, join: t.join, joinExact: t.joinExact,
          body: [], line: t.line, column: t.column,
        };
        append(node);
        stack.push({ kind: 'for', node });
        break;
      }
      case 'endfor':
        if (stack[stack.length - 1].kind !== 'for') {
          throw new TriplateSyntaxError('%endfor without a matching %for', t.line, t.column);
        }
        stack.pop();
        break;
      case 'if': {
        const node: IfNode = { type: 'if', branches: [{ cond: t.cond, body: [] }], line: t.line, column: t.column };
        append(node);
        stack.push({ kind: 'if', node, inElse: false });
        break;
      }
      case 'elif': {
        const f = stack[stack.length - 1];
        if (f.kind !== 'if' || f.inElse) throw new TriplateSyntaxError('%elif without a matching %if', t.line, t.column);
        f.node.branches.push({ cond: t.cond, body: [] });
        break;
      }
      case 'else': {
        const f = stack[stack.length - 1];
        if (f.kind !== 'if' || f.inElse) throw new TriplateSyntaxError('%else without a matching %if', t.line, t.column);
        f.inElse = true;
        f.node.elseBody = [];
        break;
      }
      case 'endif':
        if (stack[stack.length - 1].kind !== 'if') {
          throw new TriplateSyntaxError('%endif without a matching %if', t.line, t.column);
        }
        stack.pop();
        break;
    }
  }
  if (stack.length > 1) {
    const f = stack[stack.length - 1];
    const open = f.kind === 'for' ? f.node : f.kind === 'if' ? f.node : undefined;
    throw new TriplateSyntaxError('unclosed block directive', open?.line, open?.column);
  }
  return root;
}

/** Compile-time check: every reference resolves to a declared param or an in-scope loop variable. */
function validateScopes(nodes: Node[], bound: Set<string>): void {
  const checkHead = (path: RefPath, line: number, column: number) => {
    if (!bound.has(path[0])) {
      throw new TriplateSyntaxError(`undeclared variable: ${path.join('.')}`, line, column);
    }
  };
  const checkCond = (c: Cond) => checkHead(c.path, c.line, c.column);

  for (const node of nodes) {
    switch (node.type) {
      case 'value':
        checkHead(node.path, node.line, node.column);
        break;
      case 'interp':
        if (node.lang && 'path' in node.lang) checkHead(node.lang.path, node.line, node.column);
        for (const p of node.parts) if ('path' in p) checkHead(p.path, p.line, p.column);
        break;
      case 'iri':
        for (const p of node.parts) if ('path' in p) checkHead(p.path, p.line, p.column);
        break;
      case 'for': {
        checkHead(node.source, node.line, node.column);
        const inner = new Set(bound);
        inner.add(node.item);
        validateScopes(node.body, inner);
        break;
      }
      case 'if':
        for (const b of node.branches) {
          checkCond(b.cond);
          validateScopes(b.body, bound);
        }
        if (node.elseBody) validateScopes(node.elseBody, bound);
        break;
    }
  }
}

File: triplate-java/src/main/java/dev/triplate/Parser.java
package dev.triplate;

import dev.triplate.Ast.Branch;
import dev.triplate.Ast.CompiledTemplateData;
import dev.triplate.Ast.Cond;
import dev.triplate.Ast.ExampleSet;
import dev.triplate.Ast.ForNode;
import dev.triplate.Ast.IfNode;
import dev.triplate.Ast.Node;
import dev.triplate.Ast.ParamDecl;
import dev.triplate.Ast.Schema;
import dev.triplate.Ast.TypeExpr;
import dev.triplate.Lexer.Token;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;

/** Turns the token stream into the compiled template model. Port of {@code parser.ts}. */
final class Parser {
  private Parser() {}

  static CompiledTemplateData parse(String template) {
    List<Token> tokens = Lexer.tokenize(template);

    Schema schema = null;
    List<ExampleSet> examples = new ArrayList<>();
    List<Token> body = new ArrayList<>();
    for (Token t : tokens) {
      if (t instanceof Lexer.ParamsTok p) {
        if (schema != null) {
          throw new TriplateSyntaxError("duplicate params section in frontmatter", p.line(), p.column());
        }
        Map<String, TypeExpr> byName = new LinkedHashMap<>();
        for (ParamDecl d : p.decls()) {
          if (byName.containsKey(d.name())) {
            throw new TriplateSyntaxError("duplicate parameter: " + d.name(), p.line(), p.column());
          }
          byName.put(d.name(), d.type());
        }
        schema = new Schema(p.decls(), byName);
      } else if (t instanceof Lexer.ExampleTok e) {
        for (ExampleSet existing : examples) {
          if (existing.id().equals(e.id())) {
            throw new TriplateSyntaxError("duplicate example id: " + e.id(), e.line(), e.column());
          }
        }
        examples.add(new ExampleSet(e.id(), e.description(), e.bindings(), e.line(), e.column()));
      } else if (t instanceof Lexer.Txt txt && schema == null) {
        if (!txt.value().trim().isEmpty()) {
          throw new TriplateSyntaxError("content before the --- frontmatter header");
        }
      } else {
        body.add(t);
      }
    }
    if (schema == null) {
      throw new TriplateSyntaxError("missing --- frontmatter header (--- params { … } ---)");
    }

    List<Node> tree = buildTree(body);
    validateScopes(tree, new HashSet<>(schema.byName().keySet()));
    return new CompiledTemplateData(schema, examples, tree);
  }

  private static final class Frame {
    final String kind; // "root", "for", "if"
    List<Node> children;
    ForNode forNode;
    IfNode ifNode;
    boolean inElse;

    Frame(String kind) {
      this.kind = kind;
    }
  }

  private static List<Node> buildTree(List<Token> tokens) {
    List<Node> root = new ArrayList<>();
    Deque<Frame> stack = new ArrayDeque<>();
    Frame rootFrame = new Frame("root");
    rootFrame.children = root;
    stack.push(rootFrame);

    for (Token t : tokens) {
      if (t instanceof Lexer.Txt txt) {
        current(stack).add(new Ast.TextNode(txt.value()));
      } else if (t instanceof Lexer.Val v) {
        current(stack).add(new Ast.ValueNode(v.path(), v.spread(), v.join(), v.joinExact(), v.line(), v.column()));
      } else if (t instanceof Lexer.Interp i) {
        current(stack).add(new Ast.InterpNode(i.parts(), i.lang(), i.datatype(), i.line(), i.column()));
      } else if (t instanceof Lexer.Iri ir) {
        current(stack).add(new Ast.IriNode(ir.parts(), ir.line(), ir.column()));
      } else if (t instanceof Lexer.ForTok f) {
        ForNode node = new ForNode(f.item(), f.source(), f.join(), f.joinExact(), f.line(), f.column());
        current(stack).add(node);
        Frame frame = new Frame("for");
        frame.forNode = node;
        stack.push(frame);
      } else if (t instanceof Lexer.EndForTok ef) {
        if (!stack.peek().kind.equals("for")) {
          throw new TriplateSyntaxError("%endfor without a matching %for", ef.line(), ef.column());
        }
        stack.pop();
      } else if (t instanceof Lexer.IfTok ift) {
        IfNode node = new IfNode(ift.line(), ift.column());
        node.branches.add(new Branch(ift.cond()));
        current(stack).add(node);
        Frame frame = new Frame("if");
        frame.ifNode = node;
        stack.push(frame);
      } else if (t instanceof Lexer.ElifTok el) {
        Frame f = stack.peek();
        if (!f.kind.equals("if") || f.inElse) {
          throw new TriplateSyntaxError("%elif without a matching %if", el.line(), el.column());
        }
        f.ifNode.branches.add(new Branch(el.cond()));
      } else if (t instanceof Lexer.ElseTok el) {
        Frame f = stack.peek();
        if (!f.kind.equals("if") || f.inElse) {
          throw new TriplateSyntaxError("%else without a matching %if", el.line(), el.column());
        }
        f.inElse = true;
        f.ifNode.elseBody = new ArrayList<>();
      } else if (t instanceof Lexer.EndIfTok ei) {
        if (!stack.peek().kind.equals("if")) {
          throw new TriplateSyntaxError("%endif without a matching %if", ei.line(), ei.column());
        }
        stack.pop();
      }
    }
    if (stack.size() > 1) {
      Frame f = stack.peek();
      Integer line = f.kind.equals("for") ? f.forNode.line : f.kind.equals("if") ? f.ifNode.line : null;
      Integer col = f.kind.equals("for") ? f.forNode.column : f.kind.equals("if") ? f.ifNode.column : null;
      throw new TriplateSyntaxError("unclosed block directive", line, col);
    }
    return root;
  }

  private static List<Node> current(Deque<Frame> stack) {
    Frame f = stack.peek();
    if (f.kind.equals("root")) return f.children;
    if (f.kind.equals("for")) return f.forNode.body;
    return f.inElse ? f.ifNode.elseBody : f.ifNode.branches.get(f.ifNode.branches.size() - 1).body;
  }

  /** Compile-time check: every reference resolves to a declared param or an in-scope loop variable. */
  private static void validateScopes(List<Node> nodes, Set<String> bound) {
    for (Node node : nodes) {
      if (node instanceof Ast.ValueNode v) {
        checkHead(v.path(), v.line(), v.column(), bound);
      } else if (node instanceof Ast.InterpNode i) {
        if (i.lang() instanceof Ast.PathLang pl) checkHead(pl.path(), i.line(), i.column(), bound);
        for (Ast.Part p : i.parts()) {
          if (p instanceof Ast.HolePart h) checkHead(h.path(), h.line(), h.column(), bound);
        }
      } else if (node instanceof Ast.IriNode ir) {
        for (Ast.Part p : ir.parts()) {
          if (p instanceof Ast.HolePart h) checkHead(h.path(), h.line(), h.column(), bound);
        }
      } else if (node instanceof ForNode f) {
        checkHead(f.source, f.line, f.column, bound);
        Set<String> inner = new HashSet<>(bound);
        inner.add(f.item);
        validateScopes(f.body, inner);
      } else if (node instanceof IfNode iff) {
        for (Branch b : iff.branches) {
          checkCond(b.cond, bound);
          validateScopes(b.body, bound);
        }
        if (iff.elseBody != null) validateScopes(iff.elseBody, bound);
      }
    }
  }

  private static void checkHead(List<String> path, int line, int column, Set<String> bound) {
    if (!bound.contains(path.get(0))) {
      throw new TriplateSyntaxError("undeclared variable: " + String.join(".", path), line, column);
    }
  }

  private static void checkCond(Cond c, Set<String> bound) {
    checkHead(c.path(), c.line(), c.column(), bound);
  }
}
